rest/utilities_testing.go

```diff
@@ -2448,12 +2448,15 @@ func WaitAndAssertBackgroundManagerExpiredHeartbeat(t testing.TB, bm *db.Backgro
 
 // DocVersion represents a specific version of a document in an revID/HLV agnostic manner.
 type DocVersion struct {
-	RevTreeID string
-	CV        db.Version
+	RevTreeID     string                  // RevTreeID is the rev treee ID of a document, may be empty not present
+	HLV           *db.HybridLogicalVector // HLV is the hybrid logical vector of the document, may not be present
+	Mou           *db.MetadataOnlyUpdate  // Mou is the metadata only update of the document, may not be present
+	Cas           uint64                  // Cas is the cas value of the document
+	HasImplicitCV bool                    // If true, the HLV was constructed from cas@sourceID instead of from _vv, used for documents written to Couchbase Server without Sync Gateway
 }
```

Comment:
> There are some weird situations you end up in a testing scenario where you might add a document using SDK and only have a CV, or a testing scenario where you add a document via CBL 3.x and only have a revtree ID.
> RestTester put functions all use optimistic concurrency via revtreeids and we don't have access to those.

That was the point of this struct. So we have a single type that we are able to use to pass into MVCC functions that will work in all cases. Optimistic concurrency is the default behaviour of Couchbase Mobile, and I don't think we should be dropping that generally.

A subset of tests that rely on LWW seems OK, but it's not really how real-world deployments work and we should strive to keep the test APIs with MVCC in mind where we can, and allow simpler LWW fallback options where we do not want to deal with the complexity of multiple writers.

Reply:
A lot of confusion comes around the fact that documents are continually getting imported in these tests, and also by the nature of XDCR not copying over the `_sync` data structure.

Let's take a pretty simple case:

```mermaid
flowchart LR
    subgraph clusterA
        CBS1
        SG1
    end
    subgraph clusterB
        CBS2
        SG2
    end
    CBS1<-->CBS2 
```

If we have a single actor test that is doing the following write a document, then update the document on CBS1 as an SDK write, there are actually numerous states that the document will be in, and it can be challenging to debug.

- SDK write to CBS1, no `_vv`, `_mou`, cas1
- xdcr to CBS2, `_vv: { cvCAS: cas1, ver: cas1, src: cbs1}`, cas1
- import on CBS1: creates `_vv: { cvCAS: cas1, ver: cas1, src: cbs1}`,  `_mou: {cas: cas2, pCas: cas1}`, cas2
- import on CBS2: creates `_vv: { cvCAS: cas1, ver: cas1, src: cbs1}`,  `_mou: {cas: cas3, pCas: cas1}`, cas3

At this point, all documents are synchronized with a CV, but mismatched cas values. There are other scenarios at this point depending on whether import on CBS1 occurred first, then CBS2 would replicate the document with cas2, but there would still be a subsequent import.

Then, going to the second part of the test, an update of the document. Not knowing what the state of the cas, _vv, _mou mean it can be really challenging to debug where the values came from. If `doc.cas = _mou.cas` the evaluation of XDCR will be different. In this case we can see that the `_mou` looks different, but as soon as another write occurs it will be challenging to determine where it is from.

I agree that we aren't always going to want LWW, which is why I thought about pushing the extra information down to the `DocVersion` data structure.

Reply:
Addressed by creating a datastructure local to topologytest as `DocMetadata`.